File: tessoku_book/a20.py
s = input()
t = input()
n = len(s)
m = len(t)

dp = [[None] * (m+1) for _ in range(n+1)]
dp[0][0] = 0
for i in range(0, n+1):
    for j in range(0, m+1):
        if i>=1 and j >=1 and s[i-1] == t[j-1]:
            dp[i][j] = dp[i-1][j-1] + 1
        elif i>= 1 and j >= 1:
            dp[i][j] = max(dp[i-1][j] , dp[i][j-1])
        elif i>=1:
            dp[i][j] = dp[i-1][j]
        elif j>=1:
            dp[i][j] = dp[i][j-1]
            
print(dp[n][m])

File: tessoku_book/a20.cpp
#include <iostream>
#include <vector>
#include<algorithm>
using namespace std;
typedef long long ll;
#define rep0(i,n) for(int i=0;i<=n;i++)
#define rep(i,n) for(int i=1;i<=n;i++)

int main(){
    string s,t;
    cin>>s;
    cin>>t;
    int dp[2009][2009];
    dp[0][0]=0;
    rep0(i,s.size())rep0(j,t.size()){
        if(i>=1 && j>=1 && s[i-1]==t[i-1])dp[i][j]=max(dp[i-1][j],dp[i][j-1],dp[i-1][j-1]+1);
        else if(i>=1 && j>=1)dp[i][j]=max(dp[i-1][j],dp[i][j-1]);
        else if(i>=1)dp[i][j]=dp[i-1][j];
        else if(j>=1)dp[i][j]=dp[i][j-1];
    }
    cout<<dp[s.size()][t.size()]<<endl;
    return 0;
}